### PR TITLE
Add new system for model data, replacing extended states

### PR DESCRIPTION
**Pull request by tterrag1098**
Completely removes anything to do with the old IExtendedBlockState.

Adds a new system of model data/properties. This works similarly to vanilla stateholders, but much simpler.

All of the forge uses of extended states (save fluid models) have been updated to use this system, so there are a few examples of how it is meant to be used. Additionally, a test use case is provided in `ForgeMod`. This will not be included in the final version.

## Justification

The old extended state system added a lot of headaches for a system that was used comparatively little. Problems caused by extended states include

- Polluting the entire blockstate system with non-flyweight objects (breaking reference comparability)
- Introducing uncertainty of types. Similar to the problems with `MutableBlockPos`, you can never be sure if a state you have is extended or not. So your code must be littered with `if (state instanceof IExtendedBlockState) state = ((IExtendedBlockState)state).getClean()`.
- Breaking model lookups. An extension of point 1, this was a common pitfall where modders would use an extended state to look up a model object and find it missing due to equals/hashcode miss.
- Boilerplate. Adding in unlisted properties to your block took a lot of code, in places not usually concerned with rendering. This brings us to...
- Bad separation of concerns. Blockstates are generally a serverside concept. Mixing in render data muddies the waters with what blockstates should be concerned with. The new system keeps render data isolated so that it can be threadsafe and not interfere with the existing state system.
- Threadsafety issues. `getExtendedState` was called on the render worker threads, and this is not clear to many newer (or even seasoned) modders. The new system removes this limitation, allowing model data to safely be provided on-thread.

## Use

There are two main user facing hooks into this system:

### TileEntity

TileEntities can push model data to the render pipeline by simply calling `requestModelDataUpdate` (defined in `IForgeTileEntity`) at any time, and overriding `getModelData`. A simple implementation would look something like

```java
private static class Tile extends TileEntity implements ITickable {
    ...
    private int counter;

    @Override
    public void tick() {
        if (world.isRemote && counter++ == 100) {
            ModelDataManager.requestModelDataRefresh(this);
            world.markBlockRangeForRenderUpdate(getPos(), getPos());
        }
    }

    @Override
    public IModelData getModelData() {
        return new ModelDataMap.Builder().withInitial(SOME_PROP, counter < 100).build();
    }
}
```
Obviously, caching of the model data can be done for more advanced logic.

### IBakedModel

All models have the chance to add to the model data after it is (possibly) provided by the TE. This is done off-thread (in the render worker) so any work done in here must be threadsafe. This is where contextual data would be provided, by reading states from the provided `IWorldReader`. An example might look like

```java
@Override
@Nonnull
public IModelData getModelData(@Nonnull IWorldReader world, @Nonnull BlockPos pos, @Nonnull IBlockState state, @Nonnull IModelData tileData) {
    if (world.getBlockState(pos.down()).getBlock() == Blocks.AIR) {
        tileData.setData(MAGIC_PROP, !tileData.getData(MAGIC_PROP));
    }
    return tileData;
}
```

This simply inverts the (cached) value from the TE, which effectively inverts the value every time the model is drawn. Not very useful but a nice example.

### IBakedModel changes

Introduces major extensions to `IForgeBakedModel`, adding an extra `IModelData` parameter to `getQuads`. A utility interface is provided in `IDynamicBakedModel`; it is not required to be used but reduces boilerplate.

### ModelDataManager

A central location caching `IModelData` instances, this is where the `requestModelDataUpdate` call from the TE ends up. Will clear itself as soon as it detects a different world instance to the last one, indicating a clientside dimension change.

### ChunkRenderTask

The `ChunkRenderTask` holds a map of cached model data objects, and eagerly fills it from the `ModelDataManager` when it is constructed. This happens on the client thread, so it eliminates threading concerns in the TE: the creation of model data in `TileEntity.getModelData` can reference game code safely. The cache lets it be used later during rendering.

### BlockModelRenderer / BlockRendererDispatcher

Many overloads have been added to extend the parameters to accept the `IModelData` for the model being rendered. This is to avoid re-querying the model for the data for each side/layer. The model data is requested once per *draw*, period.

## What's Left

You may have noticed this is a draft PR. I'm not 100% satisfied with the `IModelData` API, and there is some cleanup of patches and test code left to do. 

It has been proposed to me to do away with the `IModelData` API entirely and simply use `Object`s as the model data, since typically their use is entirely internal. However this is not always true, for example the forge animation system. It could still be done, it would be faster performance-wise, but it may not be as elegant as the current system. Suggestions are of course welcome.

### CODE REVIEW

**Comment by Cadiboo** [comment]
Mods that can use this
- LittleTiles
- Chisel (I'm pretty sure)
- NoCubes
- CTM

File: src/main/java/net/minecraftforge/client/model/ModelDataManager.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.World;
+import net.minecraftforge.client.model.data.IModelData;
+
+public class ModelDataManager
+{
+    private static WeakReference<World> currentWorld = new WeakReference<>(null);
+    
+    private static final Map<ChunkPos, Set<BlockPos>> needModelDataRefresh = new HashMap<>();
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/client/model/ModelDataManager.java:31]: why not just synchronizedMap or concurrent map?
also this name seems a bit weird to me, maybe `dirtyPositions` or `dirtyModelData`?

Comment (tterrag1098) [reply src/main/java/net/minecraftforge/client/model/ModelDataManager.java:31]: It doesn't need to be threadsafe, it should only be accessed on the client thread. Renames will be considered, it is a bit wordy, but in the end it's an internal field.

Comment (williewillus) [reply src/main/java/net/minecraftforge/client/model/ModelDataManager.java:31]: why is it synchronized on later in this class if it doesn't need to be threadsafe?

**Comment by darkevilmac** [comment]
I think it would make more sense for model data to just be an object of any type instead of requiring it to be an implementation of ```IModelData```

It's just arbitrary data being passed to baked models, so having a whole interface and management system for it seems a bit overkill.

**Comment by tterrag1098** [comment]
> I think it would make more sense for model data to just be an object of any type instead of requiring it to be an implementation of `IModelData`

From the last section:

> It has been proposed to me to do away with the `IModelData` API entirely and simply use `Object`s as the model data, since typically their use is entirely internal. However this is not always true, for example the forge animation system. It could still be done, it would be faster performance-wise, but it may not be as elegant as the current system. Suggestions are of course welcome.

I have yet to figure out how to reconcile such a simple design with the needs of the animation system.

File: src/main/java/net/minecraftforge/client/model/data/ModelDataMap.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    public <T> T setData(ModelProperty<T> prop, T data)
+    {
+        Preconditions.checkArgument(hasProperty(prop), "Cannot set model property here as it does not exist");
+        Preconditions.checkArgument(prop.test(data), "Value is invalid for this property");
+        return (T) backingMap.put(prop, data);
+    }
+    
+    public static class Builder
+    {
+        private final Map<ModelProperty<?>, Object> defaults = new IdentityHashMap<>();
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/model/data/ModelDataMap.java:40]: Just curious, are model properties identity-based?

Comment (liach) [reply src/main/java/net/minecraftforge/client/model/data/ModelDataMap.java:40]: ah, disregard that comment above.

File: src/main/java/net/minecraftforge/client/model/data/IModelData.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package net.minecraftforge.client.model.data;
+
+public interface IModelData
+{
+    boolean hasProperty(ModelProperty<?> prop);
+    
+    <T> T getData(ModelProperty<T> prop);
+    
+    <T> T setData(ModelProperty<T> prop, T data);
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/model/data/IModelData.java:24]: Consider marking the return type as nullable.

File: src/main/java/net/minecraftforge/client/model/ModelDataManager.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import net.minecraft.client.Minecraft;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.world.World;
+import net.minecraftforge.client.model.data.IModelData;
+
+public class ModelDataManager
+{
+    private static WeakReference<World> currentWorld = new WeakReference<>(null);
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/model/ModelDataManager.java:29]: Can just do explicit `WeakReference<WorldClient>`

File: src/main/java/net/minecraftforge/client/model/ModelDataManager.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            modelDataCache.invalidateAll();
+        }
+    }
+    
+    public static void requestModelDataRefresh(TileEntity te)
+    {
+        World world = te.getWorld();
+        Preconditions.checkNotNull(world, "Tile entity world must not be null");
+        Preconditions.checkArgument(world == Minecraft.getInstance().world, "Cannot request a model data refresh for a world other than the current client world");
+        synchronized (needModelDataRefresh)
```
Comment (liach) [inline src/main/java/net/minecraftforge/client/model/ModelDataManager.java:61]: What other threads can call this method? Does the network thread call this method?

File: src/main/java/net/minecraftforge/client/model/ModelDataManager.java (hunk 1 above)
Comment (cosmicdan) [reply src/main/java/net/minecraftforge/client/model/ModelDataManager.java:31]: Just my two cents - the "need" prefix is a question word, implying a boolean (e.g. a flag). It may be just an internal field, but one should care about their fellow programmer (those in future who need to understand and maintain it) :)

Comment (williewillus) [reply src/main/java/net/minecraftforge/client/model/ModelDataManager.java:31]: my question still stands, because as it is the synchronization done is inadequate. if a field is synchronized on, *every* single use of it needs to be synchronized if it has a possibility of running in parallel.

**Comment by darkevilmac** [comment]
Is there more testing that needs to be done with this and if so is there a build I can use to test it with? I need this for most of my mods to work.

File: patches/minecraft/net/minecraft/client/renderer/chunk/ChunkRenderTask.java.patch [hunk 6]
```diff
@@ -0,0 +1,37 @@
+--- a/net/minecraft/client/renderer/chunk/ChunkRenderTask.java
++++ b/net/minecraft/client/renderer/chunk/ChunkRenderTask.java
+@@ -5,8 +5,10 @@
+ import java.util.List;
+ import java.util.concurrent.locks.ReentrantLock;
+ import net.minecraft.client.renderer.RegionRenderCacheBuilder;
++import net.minecraft.util.math.BlockPos;
+ import net.minecraftforge.api.distmarker.Dist;
+ import net.minecraftforge.api.distmarker.OnlyIn;
++import net.minecraftforge.client.model.data.EmptyModelData;
```
Comment (liach) [inline patches/minecraft/net/minecraft/client/renderer/chunk/ChunkRenderTask.java.patch]: Imports.

**Comment by tterrag1098** [comment]
You can build my branch yourself, but otherwise no. I plan to address feedback and make a final draft soon.

File: src/main/java/net/minecraftforge/client/model/ModelDataManager.java (hunk 4 above)
Comment (tterrag1098) [reply src/main/java/net/minecraftforge/client/model/ModelDataManager.java:29]: There's no need, this would just add a classloading dependency on a server-unsafe class for no reason. I don't use any methods of WorldClient.

File: src/main/java/net/minecraftforge/client/model/ModelDataManager.java (hunk 1 above)
Comment (tterrag1098) [reply src/main/java/net/minecraftforge/client/model/ModelDataManager.java:31]: You are correct, the call to `cleanCaches` was in a bad spot. None of the current calls to these methods are off-thread, but I do need to support that case.

**Comment by tterrag1098** [comment]
Ready for another round of review. Patches should be mostly finalized, the last thing to do before removing the draft status would be making the ForgeMod changes into a proper test mod.

**Review by williewillus — APPROVED** [review]
looks fine to me!

**Comment by MichaelHillcox** [comment]
@tterrag1098 any update on this? We're still waiting on this feature over at https://github.com/Direwolf20-MC/BuildingGadgets
